pkg/client/introduce/client.go

```diff
@@ -0,0 +1,172 @@
+	return c.sendProposal(Options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+	})
+}
+
+// SendRequest sends request
+// sending a request means that introducee is willing to share its invitation
+// the function receives an invitation which can be sent to the introducer
+// an invitation will not be sent if introducer responded with a public invitation
+func (c *Client) SendRequest(invitation *didexchange.Invitation, destination *service.Destination) error {
+	return c.handleOutbound(&introduce.Request{
+		Type: introduce.RequestMsgType,
+		ID:   uuid.New().String(),
+	})
+}
+
+// HandleRequest executes after receiving a request (the client does not have a public Invitation)
+func (c *Client) HandleRequest(rID string, dest1, dest2 *service.Destination) error {
+	return c.sendProposal(Options{
+		Destinations: []*service.Destination{dest1, dest2},
+		ThreadID:     rID,
+	})
+}
+
+// HandleRequestWithInvitation executes after receiving a request when the client has a public Invitation
+func (c *Client) HandleRequestWithInvitation(rID string, inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(Options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+		ThreadID:     rID,
+	})
+}
+
+func (c *Client) handleOutbound(msg interface{}) error {
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	didMsg, err := service.NewDIDCommMsg(payload)
+	if err != nil {
+		return err
+	}
+
+	return c.service.HandleOutbound(didMsg, &service.Destination{})
+}
+
+// Options keeps the information needed for sending a proposal
+type Options struct {
+	// Invitation must be set when we have a public Invitation
+	Invitation *didexchange.Invitation
+	// Destinations contain one or two elements
+	// one element - for a public Invitation, otherwise two elements
+	Destinations []*service.Destination
+	// ThreadID must be an ID from incoming Request
+	ThreadID string
+}
+
+// sendProposal sends proposal
+// the function may receive an invitation or <nil> (Options.Invitation)
+// if the invitation is not <nil> it means we have a public invitation.
+// Destinations (Options.Destinations) should have one or two values, if destinations have only one value it means
+// that it is public invitation (an invitation `Options.Invitation` is required), otherwise if destinations have two
+// values, it means we do not have a public invitation (we will wait for it from one of introducees)
+// ThreadID (Options.ThreadID) need to be used when we received Request
+// Usage:
+//  Sends public invitation to the destination
+// 	sendProposal(&Options{
+// 		Invitation: invitation,
+// 		Destinations: []*service.Destination{
+// 			destination,
+// 		},
+// 	})
+// 	Forwards an invitation from one introducee to another
+// 	sendProposal(&Options{
+// 		Destinations: []*service.Destination{
+// 			destination1,
+// 			destination1,
```

Comment:
what's the idea behind having destination1 twice?

Reply:
typo `destination2`